gitlab/_backends/protocol.py

```diff
@@ -0,0 +1,34 @@
+import abc
+import sys
+from typing import Any, Dict, Optional, Union
+
+import requests
+from requests_toolbelt.multipart.encoder import MultipartEncoder  # type: ignore
+
+if sys.version_info >= (3, 8):
+    from typing import Protocol
+else:
+    from typing_extensions import Protocol
+
+
+class BackendResponse(Protocol):
+    @abc.abstractmethod
+    def __init__(self, response: requests.Response) -> None:
```

Comment:
Has this idea been tested? I mean if we want to use `httpx` instead of `requests` how will this look?

I'm assuming that the idea is that we will be able to use `BackendResponse` and `Backend` in place of both `httpx` and `requests`. So wondering how that will look. Just wanting to make sure that this will work for that case.

Reply:
This represent the existing interface which is `requests` only based. The plan is to change the interface as you described as soon as `httpx` would be introduced. Hoping that this answers your concern.

Reply:
Let's cross that bridge when we get there, we can still change the approach if it doesn't work - otherwise if we wanted to test with httpx already, we'd have a chunky PR that would also be hard to review :sweat_smile:

Reply:
But I think if the protocol defines all the methods and attributes we might be ok. It will be a problem where httpx and requests differ, for example `iter_content` vs `iter_bytes` etc. But we can maybe have little wrappers for that.

Reply:
Just to get a bit of confidence in the direction here, I created an httpx backend locally to take this for a spin. It passes almost all functional tests except a few things:
* httpx does not coerce query params to strings, which requests seems to do, so for example Enums have to explicitly provide the value
* `response_content` is backend-specific, and shouldn't live in utils
* streamed responses dont work out of the box, since it's handled differently
* multipart is also handled differently, so it doesn't work
* I completely skirted the typing issue, that will probably be a headache.
* we cannot run unit tests on both backends unless we do some wrapping of `responses`/`respx` libraries based on the backend.

Simple requests already work though. /cc @JohnVillalovos @lmilbaum